### PR TITLE
[PropertyInfo] ConstructorExtractor which has higher priority than PhpDocExtractor and ReflectionExtractor

### PR DESCRIPTION
**Pull request by karser**
| Q             | A
| ------------- | ---
| Branch?       | 4.2
| Bug fix?      | yes
| New feature?  | no 
| BC breaks?    | hopefully no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #30053
| License       | MIT

The discussion is in previous PR #30056
In short, when using PhpDocExtractor, it ignores the constructor argument type, although `argument types from the constructor are the only types that are valid for the class instantiation`.

This PR adds a separate extractor - `ConstructorExtractor` which is called first (-999) and it attempts to extract the type from constructor only, either from PhpDoc or using reflection.
I added `getTypesFromConstructor` to `PhpDocExtractor` and `ReflectionExtractor` - they implement `ConstructorArgumentTypeExtractorInterface` interface. `ConstructorExtractor` aggregates those extractors using compiler pass.

So the flow of control looks like this:
```
PropertyInfoExtractor::getTypes:
    - ConstructorExtractor::getTypes
        - PhpDocExtractor::getTypesFromConstructor
        - ReflectionExtractor::getTypesFromConstructor
    - PhpDocExtractor::getTypes
    - ReflectionExtractor::getTypes
```

### CODE REVIEW

File: src/Symfony/Component/PropertyInfo/Extractor/PhpDocExtractor.php [hunk 1]
```diff
@@ -151,6 +151,77 @@ public function getTypes($class, $property, array $context = [])
...
+        }
+    }
+
+    /**
+     * @param DocBlock $docBlock
+     * @param string   $allowedParam
+     *
+     * @return DocBlock
+     */
+    private function filterDocBlockParams(DocBlock $docBlock, $allowedParam)
```
Comment (OskarStark) [inline src/Symfony/Component/PropertyInfo/Extractor/PhpDocExtractor.php]: You should use type hints and return types wherever possible as this is for 4.2 and new code if I am right

**Comment by jvasseur** [comment]
The problem with this solution is that by replacing the property types with the ones from the constructor we will break updating the entity with the setter (when using `object_to_populate` for example) because the type will be wrong there instead.

I think the only way to solve this problem without breaking anything would be to hold constructor argument metadata separately from property metadata. Maybe we could go with the `ConstructorArgumentTypeExtractorInterface` interface introduced in this PR but instead of using a `ConstructorExtractor` we could call the `getTypesFromConstructor` directly from the denormalizer.

**Comment by karser** [comment]
@jvasseur  
How should we approach [these controversial cases](https://github.com/symfony/symfony/pull/30056#issuecomment-461480186)?
We could restrict inferring the property type from constructor to the original case only (see #30053) where the property is private and doesn't have getter/setter.
If we did that, would it still break updating the entity with the setter?

**Comment by jvasseur** [comment]
@karser I think we souldn't infer the property type from constructor arguments at all, this would leave the property type guessing only to for properties (and setter/getter) and provide a completely separate API for getting constructor types.

I was originaly not sure about guessing property types from constructor arguments (#25605) but didn't have an example ready to be sure if there would be a problem or not, I guess now we have one.

**Comment by fabpot** [comment]
As discussed in the PR for 3.4, this is a new feature and as such it should target master.

File: src/Symfony/Component/PropertyInfo/Extractor/ConstructorArgumentTypeExtractorInterface.php [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+{
+    /**
+     * Gets types of an argument from constructor.
+     *
+     * @param string $class
+     * @param string $property
+     *
+     * @return Type[]|null
+     */
+    public function getTypesFromConstructor($class, $property);
```
Comment (OskarStark) [inline src/Symfony/Component/PropertyInfo/Extractor/ConstructorArgumentTypeExtractorInterface.php]: you should add typehints and return type

File: src/Symfony/Component/PropertyInfo/Extractor/ReflectionExtractor.php [hunk 3]
```diff
@@ -119,6 +119,50 @@ public function getTypes($class, $property, array $context = [])
...
+        return [$type];
+    }
+
+    /**
+     * @param string            $property
+     * @param \ReflectionMethod $reflectionConstructor
+     *
+     * @return \ReflectionParameter|null
+     */
+    private function getReflectionParameterFromConstructor($property, \ReflectionMethod $reflectionConstructor)
```
Comment (OskarStark) [inline src/Symfony/Component/PropertyInfo/Extractor/ReflectionExtractor.php]:
```suggestion
    private function getReflectionParameterFromConstructor(string $property, \ReflectionMethod $reflectionConstructor): ?\ReflectionParameter
```

File: src/Symfony/Component/PropertyInfo/Tests/Extractor/ConstructorExtractorTest.php [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    }
+
+    public function testInstanceOf()
+    {
+        $this->assertInstanceOf('Symfony\Component\PropertyInfo\PropertyTypeExtractorInterface', $this->extractor);
+    }
+
+    public function testGetTypes()
+    {
+        $this->assertEquals([new Type(Type::BUILTIN_TYPE_STRING)], $this->extractor->getTypes('Foo', 'bar', []));
```
Comment (OskarStark) [inline src/Symfony/Component/PropertyInfo/Tests/Extractor/ConstructorExtractorTest.php]: Can you use `assertSame()` here?

File: src/Symfony/Component/PropertyInfo/Tests/Fixtures/ConstructorDummy.php [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+    /** @var int */
+    private $dateTime;
+
+    /**
+     * @param \DateTimeZone      $timezone
+     * @param int                $date       Timestamp
+     * @param \DateTimeInterface $dateObject
+     */
+    public function __construct(\DateTimeZone $timezone, $date, $dateObject, \DateTime $dateTime)
```
Comment (OskarStark) [inline src/Symfony/Component/PropertyInfo/Tests/Fixtures/ConstructorDummy.php]:
```suggestion
    public function __construct(\DateTimeZone $timezone, int $date, \DateTimeInterface $dateObject, \DateTime $dateTime)
```

Comment (OskarStark) [reply src/Symfony/Component/PropertyInfo/Tests/Fixtures/ConstructorDummy.php]: but `$dateObject` isn't used´ .... 🤔

Comment (karser) [reply src/Symfony/Component/PropertyInfo/Tests/Fixtures/ConstructorDummy.php]: @OskarStark  That's on purpose, they are used for testing phpdoc types extraction from constructor even if no such property exists.

File: src/Symfony/Component/PropertyInfo/Tests/Extractor/ConstructorExtractorTest.php (hunk 4 above)
Comment (karser) [reply src/Symfony/Component/PropertyInfo/Tests/Extractor/ConstructorExtractorTest.php]: I can't, assertSame `asserts that two variables reference the same object.`
image

File: src/Symfony/Component/PropertyInfo/Extractor/PhpDocExtractor.php (hunk 1 above)
Comment (karser) [reply src/Symfony/Component/PropertyInfo/Extractor/PhpDocExtractor.php]: @OskarStark I applied the type hints in #30335

**Comment by karser** [comment]
closing this in favor of #30335

**Pull request closed without merging**